quantum/matrix.c

```diff
@@ -15,300 +15,209 @@ GNU General Public License for more details.
-            }
+    if (debouncing_delay >= 0) {
+        dprintf("Debouncing delay remaining: %X\n", debouncing_delay);
+        if (--debouncing_delay >= 0) {
+            wait_ms(1);
         }
-    }
-#else
-    for (uint8_t i = 0; i < MATRIX_COLS; i++) {
-        select_row(i);
-        _delay_us(30);  // without this wait read unstable value.
-        matrix_row_t rows = read_cols();
-        if (matrix_reversed_debouncing[i] != rows) {
-            matrix_reversed_debouncing[i] = rows;
-            if (debouncing) {
-                debug("bounce!: "); debug_hex(debouncing); debug("\n");
+        else {
+            for (int8_t r = MATRIX_ROWS - 1; r >= 0; --r) {
+                matrix[r] = debouncing_matrix[r];
             }
-            debouncing = DEBOUNCE;
         }
-        unselect_rows();
     }
-
-    if (debouncing) {
-        if (--debouncing) {
-            _delay_ms(1);
-        } else {
-            for (uint8_t i = 0; i < MATRIX_COLS; i++) {
-                matrix_reversed[i] = matrix_reversed_debouncing[i];
-            }
-        }
-    }
-    for (uint8_t y = 0; y < MATRIX_ROWS; y++) {
-        matrix_row_t row = 0;
-        for (uint8_t x = 0; x < MATRIX_COLS; x++) {
-            row |= ((matrix_reversed[x] & (1<<y)) >> y) << x;
-        }
-        matrix[y] = row;
-    }
-#endif
-
     matrix_scan_quantum();
-
     return 1;
 }
 
-bool matrix_is_modified(void)
-{
-    if (debouncing) return false;
-    return true;
+static void toggle_row(uint8_t row) {
+    /* PINxn */
+    _SFR_IO8(row_pins[row].input_addr) = _BV(row_pins[row].bit);
 }
 
-inline
-bool matrix_is_on(uint8_t row, uint8_t col)
-{
-    return (matrix[row] & ((matrix_row_t)1<col));
+static matrix_row_t read_cols(void) {
+    matrix_row_t state = 0;
+    for (int8_t c = MATRIX_COLS - 1; c >= 0; --c) {
+        /* PINxn */
+        if (!(_SFR_IO8(col_pins[c].input_addr) & _BV(col_pins[c].bit))) {
+            state |= (matrix_row_t)1 << c;
+        }
+    }
+    return state;
 }
 
-inline
-matrix_row_t matrix_get_row(uint8_t row)
-{
+matrix_row_t matrix_get_row(uint8_t row) {
     return matrix[row];
 }
 
-void matrix_print(void)
-{
-    print("\nr/c 0123456789ABCDEF\n");
-    for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
-        phex(row); print(": ");
-        pbin_reverse16(matrix_get_row(row));
-        print("\n");
+#else
+uint8_t matrix_scan(void) {
+    for (int8_t c = MATRIX_COLS - 1; c >= 0; --c) {
+        toggle_col(c);
+        matrix_col_t state = read_rows();
+        if (debouncing_matrix[c] != state) {
+            debouncing_matrix[c] = state;
+            debouncing_delay = DEBOUNCING_DELAY;
+        }
+        toggle_col(c);
     }
-}
-
-uint8_t matrix_key_count(void)
-{
-    uint8_t count = 0;
-    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
-        count += bitpop16(matrix[i]);
+    if (debouncing_delay >= 0) {
+        dprintf("Debouncing delay remaining: %X\n", debouncing_delay);
+        if (--debouncing_delay >= 0) {
```

Comment:
Sorry, GitHub has been acting up and I'm just seeing this. I didn't write the original condition, but I get what you're saying. I do strive to make my code clean enough not to require comments, though.